ocaml/xapi/taskHelper.ml

```diff
@@ -217,58 +235,73 @@ let is_cancelling ~__context =
-  assert_op_valid ~__context self ;
-  let status = Db_actions.DB_Action.Task.get_status ~__context ~self in
-  if status = `pending then (
-    Db_actions.DB_Action.Task.set_progress ~__context ~self ~value:1. ;
-    Db_actions.DB_Action.Task.set_finished ~__context ~self
-      ~value:(Date.of_float (Unix.time ())) ;
-    Db_actions.DB_Action.Task.set_status ~__context ~self ~value:`cancelled ;
-    Db_actions.DB_Action.Task.set_allowed_operations ~__context ~self ~value:[]
-  ) else
-    debug "the status of %s is %s; cannot set it to `cancelled"
-      (Ref.really_pretty_and_small self)
-      (status_to_string status)
-
-let cancel ~__context =
-  operate_on_db_task ~__context (fun self -> cancel_this ~__context ~self)
-
-let failed ~__context exn =
-  let backtrace = Printexc.get_backtrace () in
-  Context.complete_tracing __context ~error:(exn, backtrace) ;
-  let code, params = ExnHelper.error_of_exn exn in
-  operate_on_db_task ~__context (fun self ->
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () ->
+      assert_op_valid ~__context self ;
       let status = Db_actions.DB_Action.Task.get_status ~__context ~self in
       if status = `pending then (
         Db_actions.DB_Action.Task.set_progress ~__context ~self ~value:1. ;
-        Db_actions.DB_Action.Task.set_error_info ~__context ~self
-          ~value:(code :: params) ;
-        Db_actions.DB_Action.Task.set_backtrace ~__context ~self
-          ~value:(Sexplib.Sexp.to_string Backtrace.(sexp_of_t (get exn))) ;
         Db_actions.DB_Action.Task.set_finished ~__context ~self
           ~value:(Date.of_float (Unix.time ())) ;
+        Db_actions.DB_Action.Task.set_status ~__context ~self ~value:`cancelled ;
         Db_actions.DB_Action.Task.set_allowed_operations ~__context ~self
-          ~value:[] ;
-        if code = Api_errors.task_cancelled then
-          Db_actions.DB_Action.Task.set_status ~__context ~self
-            ~value:`cancelled
-        else
-          Db_actions.DB_Action.Task.set_status ~__context ~self ~value:`failure
+          ~value:[]
       ) else
-        debug "the status of %s is %s; cannot set it to %s"
+        debug "the status of %s is %s; cannot set it to `cancelled"
           (Ref.really_pretty_and_small self)
           (status_to_string status)
-          (if code = Api_errors.task_cancelled then "`cancelled" else "`failure")
-  )
+    )
+    (fun () -> Context.complete_tracing __context)
+
+let cancel ~__context =
+  Context.with_tracing ~__context __FUNCTION__ @@ fun __context ->
+  operate_on_db_task ~__context (fun self -> cancel_this ~__context ~self)
+
+let failed ~__context exn =
+  let backtrace = Printexc.get_backtrace () in
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () ->
+      let code, params = ExnHelper.error_of_exn exn in
+      operate_on_db_task ~__context (fun self ->
+          let status = Db_actions.DB_Action.Task.get_status ~__context ~self in
+          if status = `pending then (
```

Comment:
I would prefer less nesting by using a `match` over `status`. When cancellation can be handled using a `when` condition.

Reply:
I didn't use the `when` condition but I addressed this here https://github.com/xapi-project/xen-api/pull/5735/commits/bd8aec6d8303c4318d7f2fae167aba045dbb52c3.